src/dtls13.c

```diff
@@ -0,0 +1,996 @@
+ * @ssl: [in] ssl object
+ * @input: [in] buffer where the header is
+ * @input_size: [in] size of the input buffer
+ * @record_length: [out] record length
+ * @header_length: [out] record header length
+ * @seq_lo: [out] lower byte of seq number
+ * @seq_hi: [out] higher byte of seq number (not always present)
+ * @seq_hi_present: [out] if 1 then the higher byte is present in the header
+ * @epoch_bits: [out] lower two bits of the epoch
+ *
+ * Further info: draft43 section 4
+ *
+ * return 0 on success
+ */
+int dtls13_parse_unifed_record_layer(WOLFSSL *ssl,
+                                     const byte *input,
+                                     word16 input_size,
+                                     word16 *record_length,
+                                     word16 *header_length,
+                                     byte *seq_lo,
+                                     byte *seq_hi,
+                                     int *seq_hi_present,
+                                     byte *epoch_bits)
+{
+    byte seq_len, has_length;
+    word16 length;
+    byte *seq_num;
+    int ret;
+
+    if (input == NULL || input_size == 0)
+        return BAD_FUNC_ARG;
+
+    if (*input & C_BIT) {
+        WOLFSSL_MSG("DTLS1.3 header with connection ID. Not supported");
+        return WOLFSSL_NOT_IMPLEMENTED;
+    }
+
+    length = HEADER_FLAGS_SIZE;
+
+    seq_len = (*input & S_BIT) != 0 ? SEQ_16_LEN : SEQ_8_LEN;
+    has_length = *input & L_BIT;
+    *epoch_bits = *input & EE_MASK;
+
+    length += seq_len;
+
+    if (input_size < length)
+            return BUFFER_ERROR;
+
+    if (has_length) {
+        if (input_size < length + LEN_SIZE)
+            return BUFFER_ERROR;
+
+        ato16(input + length, record_length);
+        length += LEN_SIZE;
+
+        /* DTLS message must fit inside a datagram  */
+        if (input_size < length + *record_length)
+            return LENGTH_ERROR;
+
+    } else {
+        /* length not present. The size of the record is the all the remaining
+           data received with this datagram */
+        *record_length = input_size - length;
+    }
+
+    /* minimum size for a dtls1.3 packet is 16 bytes (to have enough ciphertext
+       to create record number xor mask). (draft 43 - Sec 4.2.3) */
+    if (*record_length < RN_MASK_SIZE)
+            return LENGTH_ERROR;
+
+    seq_num = (byte*)(input + HEADER_FLAGS_SIZE);
+    /* cast is needed to override the const qualifier */
+    ret = dtls13_encrypt_decrypt_record_number(ssl, seq_num, seq_len,
+                                               (byte *)input + length,
+                                               DEPROTECT);
+    if (ret != 0)
+        return ret;
+
+    *header_length = length;
+
+    if (seq_len == SEQ_16_LEN) {
+        *seq_hi_present =  1;
+        *seq_hi = seq_num[0];
+        *seq_lo = seq_num[1];
+    } else {
+        *seq_hi_present = 0;
+        *seq_hi = seq_num[0];
```

Comment:
Shouldn't this be setting `seq_lo` instead?

Reply:
nice catch, thanks